packages/ui/src/components/TagList/index.tsx

```diff
@@ -82,77 +107,197 @@ export const TagList = ({
+  const memoizedResult = useMemo(() => {
+    let tmpThreshold = threshold
+    if (
+      tags.length > 0 &&
+      tags
+        .slice(0, tmpThreshold)
+        .reduce<string>((acc, tag) => acc + getTagLabel(tag), '').length >
+        maxLength
+    ) {
+      tmpThreshold -= 1
+    }
+
+    const potentiallyVisibleTagsLength =
+      tags.length > tmpThreshold || false ? tmpThreshold : tags.length
+    const potentiallyVisibleTags = tags.slice(0, potentiallyVisibleTagsLength)
+    const surelyHiddenTags = tags.slice(potentiallyVisibleTagsLength)
+
+    return {
+      tmpThreshold,
+      potentiallyVisibleTags,
+      surelyHiddenTags,
+    }
+  }, [maxLength, tags, threshold])
+
+  const { tmpThreshold, potentiallyVisibleTags, surelyHiddenTags } =
+    memoizedResult
+
+  // compute visible tags and hidden ones based on the container width and
+  // what can fit into it from the potentially visible tags
+  useEffect(() => {
+    if (!tags.length || !containerRef.current || !measureRef.current) {
+      return
+    }
+
+    if (multiline) {
+      setVisibleTags(potentiallyVisibleTags)
+      setHiddenTags(surelyHiddenTags)
+
+      return
+    }
+
+    const parentWidth = containerRef.current.parentElement?.clientWidth || 0
+    let accumulatedWidth = 0
+    const measuredVisibleTags = []
+    const measuredHiddenTags = []
+
+    const measureElements = measureRef.current.children[0].children
+    for (let i = 0; i < measureElements.length; i += 1) {
+      accumulatedWidth +=
+        (measureElements[i] as HTMLDivElement).offsetWidth +
+        parseInt(TAGS_GAP, 10)
+      if (accumulatedWidth <= parentWidth) {
+        measuredVisibleTags.push(tags[i])
+      } else {
+        measuredHiddenTags.push(tags[i])
+      }
+    }
 
-  const [isVisible, setIsVisible] = useState(false)
+    setVisibleTags(measuredVisibleTags)
+    setHiddenTags(measuredHiddenTags.concat(surelyHiddenTags))
+  }, [multiline, potentiallyVisibleTags, surelyHiddenTags, tags, tmpThreshold])
+
+  // get the popover trigger ref when we show it to be able to:
+  // - add a first tag with ellipsis when the first tag does not fit in the parent container
+  // - adjust the size of the first tag we do this only when we show one tag and we have more hidden ones
+  useEffect(() => {
+    if (popoverTriggerRef.current?.clientWidth) {
+      const newPopoverTriggerWidth = popoverTriggerRef.current.clientWidth
+
+      // a check to know if we need to ellipsis the first tag when needed
+      if (visibleTags.length === 1 && hiddenTags.length > 0) {
+        setPopoverTriggerWidth(newPopoverTriggerWidth)
+      }
+
+      // add a first tag with ellipsis when the first tag does not fit in the parent container
+      if (visibleTags.length === 0 && hiddenTags.length > 0) {
+        const visibleTagsCopy = [...visibleTags]
+        const hiddenTagsCopy = [...hiddenTags]
+
+        const tagToMove = hiddenTagsCopy.pop()
+        visibleTagsCopy.push(tagToMove as TagType)
```

Comment:
Same here we should avoid mutation as much as possible, you will have to create a new array:

```tsx
const newArray = [...visibleTagsCopy, tagToMove]
```